### PR TITLE
Fix federated user scope validation

### PR DESCRIPTION
**Pull request by SujanSanjula96**
### Proposed changes in this pull request

Fixes https://github.com/wso2/product-is/issues/18080

### CODE REVIEW

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/approles/impl/AppAssociatedRolesResolverImpl.java [hunk 1]
```diff
@@ -154,14 +155,58 @@ private Set<String> getAllRolesOfFederatedUser(AuthenticatedUser authenticatedUs
...
         String tenantDomain = authenticatedUser.getTenantDomain();
-        List<String> userIDPGroups = getFederatedUserIDPGroup(authenticatedUser);
-        if (CollectionUtils.isEmpty(userIDPGroups)) {
+        String idpName = authenticatedUser.getFederatedIdPName();
+        IdentityProvider identityProvider = getIDP(idpName, tenantDomain);
+        if (identityProvider == null) {
+            return Collections.emptySet();
+        }
+        IdPGroup[] idpGroups = identityProvider.getIdPGroupConfig();
+        List<String> userIdPGroups = getFederatedUserIDPGroups(authenticatedUser, identityProvider);
```
Comment (AnuradhaSK) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/approles/impl/AppAssociatedRolesResolverImpl.java:164]:
```suggestion
        List<String> userIdPGroups = getIdpGroupNamesOfFederatedUser(authenticatedUser, identityProvider);
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/approles/impl/AppAssociatedRolesResolverImpl.java [hunk 2]
```diff
@@ -154,14 +155,58 @@ private Set<String> getAllRolesOfFederatedUser(AuthenticatedUser authenticatedUs
...
+    }
+
+    /**
+     * Get user IDP group ids.
+     *
+     * @param userIdPGroups User IDP group names.
+     * @param idpGroups     IDP groups.
+     * @return User IDP group ids.
+     */
+    private List<String> getUserIdPGroupIdList(List<String> userIdPGroups, IdPGroup[] idpGroups) {
```
Comment (AnuradhaSK) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/approles/impl/AppAssociatedRolesResolverImpl.java:199]:
```suggestion
    private List<String> getIdPGroupIdListOfUser(List<String> userIdPGroups, IdPGroup[] idpGroups) {
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/approles/impl/AppAssociatedRolesResolverImpl.java [hunk 3]
```diff
@@ -154,14 +155,58 @@ private Set<String> getAllRolesOfFederatedUser(AuthenticatedUser authenticatedUs
...
+            return Collections.emptySet();
+        }
+        IdPGroup[] idpGroups = identityProvider.getIdPGroupConfig();
+        List<String> userIdPGroups = getFederatedUserIDPGroups(authenticatedUser, identityProvider);
+        if (CollectionUtils.isEmpty(userIdPGroups)) {
             return Collections.emptySet();
         }
-        List<String> roleIdsFromIDPGroups = getRoleIdsOfIdpGroups(userIDPGroups, tenantDomain);
+        List<String> userIdPGroupIdList = getUserIdPGroupIdList(userIdPGroups, idpGroups);
+        List<String> roleIdsFromIDPGroups = getRoleIdsOfIdpGroups(userIdPGroupIdList, tenantDomain);
```
Comment (AnuradhaSK) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/approles/impl/AppAssociatedRolesResolverImpl.java:169]:
```suggestion
        List<String> roleIdsFromIdPGroups = getRoleIdsOfIdpGroups(userIdPGroupIdList, tenantDomain);
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/approles/impl/AppAssociatedRolesResolverImpl.java (hunk 1 above)
Comment (SujanSanjula96) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/approles/impl/AppAssociatedRolesResolverImpl.java]: Fixed with https://github.com/wso2/carbon-identity-framework/pull/5219/commits/b9e9242e76204067e4b71ea8c7989e85fac6a5f0

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/approles/impl/AppAssociatedRolesResolverImpl.java (hunk 2 above)
Comment (SujanSanjula96) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/approles/impl/AppAssociatedRolesResolverImpl.java]: Fixed with https://github.com/wso2/carbon-identity-framework/pull/5219/commits/b9e9242e76204067e4b71ea8c7989e85fac6a5f0

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/approles/impl/AppAssociatedRolesResolverImpl.java (hunk 3 above)
Comment (SujanSanjula96) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/approles/impl/AppAssociatedRolesResolverImpl.java]: Fixed with https://github.com/wso2/carbon-identity-framework/pull/5219/commits/b9e9242e76204067e4b71ea8c7989e85fac6a5f0

**Review by AnuradhaSK — APPROVED** [review]

**Pull request merged**